pkg/proxy/service_test.go

```diff
@@ -0,0 +1,534 @@
+}
+
+func makeTestServiceInfo(clusterIP string, port int, protocol string, healthcheckNodePort int) *fakeServiceInfo {
+	info := &fakeServiceInfo{
+		clusterIP: net.ParseIP(clusterIP),
+		port:      port,
+		protocol:  api.Protocol(protocol),
+	}
+	if healthcheckNodePort != 0 {
+		info.healthCheckNodePort = healthcheckNodePort
+	}
+	return info
+}
+
+func newFakeServiceInfo(svcPortName ServicePortName, servicePort *api.ServicePort, service *api.Service) ServicePort {
+	info := &fakeServiceInfo{
+		clusterIP: net.ParseIP(service.Spec.ClusterIP),
+		port:      int(servicePort.Port),
+		protocol:  servicePort.Protocol,
+	}
+	if apiservice.NeedsHealthCheck(service) {
+		p := service.Spec.HealthCheckNodePort
+		if p != 0 {
+			info.healthCheckNodePort = int(p)
+		}
+	}
+	return info
+}
+
+func makeTestService(namespace, name string, svcFunc func(*api.Service)) *api.Service {
+	svc := &api.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Annotations: map[string]string{},
+		},
+		Spec:   api.ServiceSpec{},
+		Status: api.ServiceStatus{},
+	}
+	svcFunc(svc)
+	return svc
+}
+
+func addTestPort(array []api.ServicePort, name string, protocol api.Protocol, port, nodeport int32, targetPort int) []api.ServicePort {
+	svcPort := api.ServicePort{
+		Name:       name,
+		Protocol:   protocol,
+		Port:       port,
+		NodePort:   nodeport,
+		TargetPort: intstr.FromInt(targetPort),
+	}
+	return append(array, svcPort)
+}
+
+func makeNSN(namespace, name string) types.NamespacedName {
+	return types.NamespacedName{Namespace: namespace, Name: name}
+}
+
+func makeServicePortName(ns, name, port string) ServicePortName {
+	return ServicePortName{
+		NamespacedName: makeNSN(ns, name),
+		Port:           port,
+	}
+}
+
+func Test_serviceToServiceMap(t *testing.T) {
+	testCases := []struct {
+		service  *api.Service
+		expected map[ServicePortName]*fakeServiceInfo
+	}{
+		{
+			// Case[0]: nothing
+			service:  nil,
+			expected: map[ServicePortName]*fakeServiceInfo{},
+		},
+		{
+			// Case[1]: headless service
+			service: makeTestService("somewhere-else", "headless", func(svc *api.Service) {
+				svc.Spec.Type = api.ServiceTypeClusterIP
+				svc.Spec.ClusterIP = api.ClusterIPNone
+				svc.Spec.Ports = addTestPort(svc.Spec.Ports, "rpc", "UDP", 1234, 0, 0)
+			}),
+			expected: map[ServicePortName]*fakeServiceInfo{},
+		},
+		{
+			// Case[2]: headless service without port
+			service: makeTestService("somewhere-else", "headless-without-port", func(svc *api.Service) {
+				svc.Spec.Type = api.ServiceTypeClusterIP
+				svc.Spec.ClusterIP = api.ClusterIPNone
+			}),
+			expected: map[ServicePortName]*fakeServiceInfo{},
+		},
+		{
+			// Case[3]: cluster ip service
+			service: makeTestService("somewhere-else", "cluster-ip", func(svc *api.Service) {
+				svc.Spec.Type = api.ServiceTypeClusterIP
+				svc.Spec.ClusterIP = "172.16.55.4"
+				svc.Spec.Ports = addTestPort(svc.Spec.Ports, "something", "UDP", 1234, 4321, 0)
```

Comment:
Same idea as above. Try giving it a name like "test-port" or "port1" or something like that just to clearly indicate what it is.